File: backend-rust/tests/support/tenant_assertions.rs
use std::collections::BTreeMap;

use serde_json::Value;

use super::tenant_fixture::{
    InternalRequest, QueueMessage, Resource, SseEvent, REQUIRED_INTERNAL_HEADERS,
};

pub fn assert_same_tenant(parent: &Resource, child: &Resource) {
    assert_eq!(parent.tenant_id, child.tenant_id);
    assert_eq!(child.parent_id.as_deref(), Some(parent.id.as_str()));
}

pub fn assert_tenant_scoped(expected_tenant_id: &str, resources: &[Resource]) {
    assert!(
        resources
            .iter()
            .all(|resource| resource.tenant_id == expected_tenant_id),
        "resource collection crossed tenant boundary: expected {expected_tenant_id}, got {:?}",
        resources
            .iter()
            .map(|resource| resource.tenant_id.as_str())
            .collect::<Vec<_>>()
    );
}

pub fn assert_cross_tenant(requester_tenant_id: &str, target: &Resource) {
    assert_ne!(requester_tenant_id, target.tenant_id);
}

pub fn assert_tenant_path(tenant_id: &str, path: &str) {
    assert!(
        path.contains(&format!("/{tenant_id}/")),
        "path is not tenant-prefixed: {path}"
    );
}

pub fn assert_sse_event(tenant_id: &str, event: &SseEvent) {
    assert_eq!(event.schema_version, 1);
    assert!(!event.event_id.is_empty());
    assert!(!event.event.is_empty());
    assert_eq!(event.tenant_id, tenant_id);
    assert!(!event.task_id.is_empty());
    assert!(event.occurred_at.ends_with('Z'));
    assert!(event.data.is_object());
}

pub fn assert_queue_message(
    tenant_id: &str,
    actor_user_id: &str,
    message: &QueueMessage,
) {
    assert_eq!(message.schema_version, 1);
    assert_eq!(message.tenant_id, tenant_id);
    assert_eq!(message.actor_user_id, actor_user_id);
    assert!(!message.task_id.is_empty());
    assert!(!message.request_id.is_empty());

    let json = message.as_json();
    for field in ["schema_version", "tenant_id", "task_id", "actor_user_id", "request_id"] {
        assert!(json.get(field).is_some(), "queue field missing: {field}");
    }
}

pub fn assert_internal_request(tenant_id: &str, request: &InternalRequest) {
    assert_eq!(request.tenant_id, tenant_id);
    assert!(!request.user_id.is_empty());
    assert!(!request.request_id.is_empty());
    assert_eq!(request.body_sha256.len(), 64);
    assert!(request.body_sha256.chars().all(is_lower_hex));
    assert!(!request.canonical_request.ends_with('\n'));
    assert!(request.canonical_request.contains(&format!("\n{tenant_id}\n")));
    assert_eq!(
        request.replay_key,
        format!("{}:{}", request.tenant_id, request.request_id)
    );
    for header in REQUIRED_INTERNAL_HEADERS {
        assert!(
            request.headers.contains_key(header),
            "required internal header missing: {header}"
        );
    }
    assert_eq!(request.headers.get("X-Tenant-Id"), Some(&request.tenant_id));
    assert_eq!(request.headers.get("X-User-Id"), Some(&request.user_id));
    assert_eq!(request.headers.get("X-Request-Id"), Some(&request.request_id));
    assert_eq!(
        request.headers.get("X-Internal-Timestamp"),
        Some(&request.timestamp)
    );
    assert_eq!(
        request.headers.get("X-Internal-Signature"),
        Some(&request.signature)
    );
    assert_signature_shape(&request.signature);
}

pub fn assert_signature_shape(signature: &str) {
    assert!(signature.starts_with("v1="));
    let digest = &signature[3..];
    assert_eq!(digest.len(), 64);
    assert!(digest.chars().all(is_lower_hex));
}

pub fn assert_required_headers(headers: &BTreeMap<String, String>) {
    for header in REQUIRED_INTERNAL_HEADERS {
        assert!(headers.contains_key(header), "missing header: {header}");
    }
}

fn is_lower_hex(value: char) -> bool {
    value.is_ascii_digit() || ('a'..='f').contains(&value)
}

pub fn assert_error_code(data: &Value, expected: &str) {
    assert_eq!(data.get("error_code").and_then(Value::as_str), Some(expected));
}

File: backend-java/src/test/java/com/ithsd/smart_tender/tenant/fixture/TenantAssertions.java
package com.ithsd.smart_tender.tenant.fixture;

import org.assertj.core.api.Assertions;

import java.util.Collection;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Assertions shared by contract-spec tests and future endpoint tests.
 */
public final class TenantAssertions {

    private TenantAssertions() {
    }

    public static void assertSameTenant(
            TenantFixture.Resource parent,
            TenantFixture.Resource child
    ) {
        assertThat(parent.tenantId()).isEqualTo(child.tenantId());
        assertThat(child.parentId()).isEqualTo(parent.id());
    }

    public static void assertTenantScoped(
            String expectedTenantId,
            TenantFixture.Resource resource
    ) {
        assertThat(resource.tenantId()).isEqualTo(expectedTenantId);
    }

    public static void assertTenantScoped(
            String expectedTenantId,
            Collection<? extends TenantFixture.Resource> resources
    ) {
        assertThat(resources)
                .allSatisfy(resource -> assertThat(resource.tenantId()).isEqualTo(expectedTenantId));
    }

    public static void assertCrossTenant(
            TenantFixture.Tenant requester,
            TenantFixture.Resource target
    ) {
        assertThat(target.tenantId()).isNotEqualTo(requester.id());
    }

    public static void assertTenantPath(
            String tenantId,
            String path
    ) {
        assertThat(path).contains("/" + tenantId + "/");
    }

    public static void assertSseEvent(
            TenantFixture.Tenant tenant,
            TenantFixture.SseEvent event
    ) {
        assertThat(event.schemaVersion()).isEqualTo(1);
        assertThat(event.eventId()).isNotBlank();
        assertThat(event.event()).isNotBlank();
        assertThat(event.tenantId()).isEqualTo(tenant.id());
        assertThat(event.taskId()).isNotBlank();
        assertThat(event.occurredAt()).endsWith("Z");
        assertThat(event.data()).isNotEmpty();
    }

    public static void assertQueueMessage(
            TenantFixture.Tenant tenant,
            TenantFixture.QueueMessage message,
            String expectedActorUserId
    ) {
        assertThat(message.schemaVersion()).isEqualTo(1);
        assertThat(message.tenantId()).isEqualTo(tenant.id());
        assertThat(message.taskId()).isNotBlank();
        assertThat(message.actorUserId()).isEqualTo(expectedActorUserId);
        assertThat(message.requestId()).isNotBlank();
    }

    public static void assertInternalRequest(
            TenantFixture.Tenant tenant,
            TenantFixture.InternalRequest request
    ) {
        assertThat(request.tenantId()).isEqualTo(tenant.id());
        assertThat(request.userId()).isNotBlank();
        assertThat(request.requestId()).isNotBlank();
        assertThat(request.bodySha256()).matches("[0-9a-f]{64}");
        assertThat(request.canonicalRequest()).doesNotEndWith("\n");
        assertThat(request.canonicalRequest()).contains("\n" + tenant.id() + "\n");
        assertThat(request.replayKey()).isEqualTo(tenant.id() + ":" + request.requestId());
        assertThat(request.headers()).containsEntry("X-Tenant-Id", tenant.id());
        assertThat(request.headers()).containsEntry("X-User-Id", request.userId());
        assertThat(request.headers()).containsEntry("X-Request-Id", request.requestId());
        assertThat(request.headers()).containsEntry("X-Internal-Timestamp", request.timestamp());
        assertThat(request.headers()).containsEntry("X-Internal-Signature", request.signature());
        assertThat(request.signature()).matches("v1=[0-9a-f]{64}");
    }

    public static void assertErrorShape(
            Map<String, Object> data,
            String expectedErrorCode,
            String expectedRequestId
    ) {
        assertThat(data).containsEntry("error_code", expectedErrorCode);
        assertThat(data).containsEntry("request_id", expectedRequestId);
    }
}
